### PR TITLE
task(settings): Port relier models from content-server

### PR DESCRIPTION
**Pull request by dschom**
## Because
- We are migrating content server to react

## This pull request
- Ports relier models
- Adds ability to bind models to generic context via an @Bind property decorator
- Adds ability to validate field state via a set of validators that can be specified on the @Bind decorator
- Provides implementation to bind a model to url search query string context
- Provides implementation to bind a model to url hash state context
- Provides implementation to bind a model to a storage context
- Adds a creational factory pattern for initializing relier models and their state
- Adds relier factory to app context
- Adds a useRelier hook so that we can easily access relier models in components

## Issue that this pull request solves

Closes: FXA-6437

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: packages/fxa-settings/src/models/reliers/browser-relier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/browser-relier.ts:20]: Ref lib/context/bind-context.ts for implementation.

File: packages/fxa-settings/src/models/reliers/browser-relier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/browser-relier.ts:9]: This pattern of having a 'data' interface is used across all relier models.

I am thinking these are what our controls use instead of the classes. This should make testing easy, cause you can just passing whatever state is needed. Thereby decoupling the tests that binding works as expected from the tests on the components themselves.

File: packages/fxa-settings/src/models/reliers/options/oauth-relier-options.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/options/oauth-relier-options.ts:15]: I tried to figure which parameters were required and which ones were optional. The types should reflect this.

File: packages/fxa-settings/src/models/reliers/browser-relier.ts
Comment (dschom) [reply packages/fxa-settings/src/models/reliers/browser-relier.ts:20]: The bind enforces the validation and thus acts a type gaurd.

File: packages/fxa-settings/src/models/reliers/oauth-relier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/oauth-relier.ts:25]: The T.snakeCase lets the binding know that the context holds this value under a key with snake case, i.e. keys_jkw.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/relier-factory.ts:38]: This is intended to take the logic of creating state out of the models themselves. In theory this should making testing easier, because now we've shifted the responsibility of creation to this factory, which means we can test creation in isolation. Also, when we need to use the models in our components there is no longer any assumption about how to create state in the model itself, which should making mocking easier / more direct.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/relier-factory.ts:81]: Most the decision making for the relier factory actually happens here. This is a bit unconventional, but it was the easiest way to port the existing logic over without having to change much. For example a reference to `this.__isVerification()` is easily retained by making this its own class.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/relier-factory.ts:174]: I'd like to understand better why we need to do this. It required pulling the storageContext. It'd be great if the code wasn't as concerned with this. I am sure it's needed for several things though. Perhaps, the model binding that was created for reliers could also be created for important session context state.

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/context/implementations/storage-context.ts:26]: This was ported from content server. I am not in love with the way this works. I think the JSON serialization adds unnecessary overhead, and we could just use storage like a keystore which would be more performant, more direct, and possibly not require concerns about load and persist. 

At this point, we have to maintain backwards compatibility, but it might be interesting to revisit this.

File: packages/fxa-settings/src/lib/context/implementations/url-context.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/context/implementations/url-context.ts:34]: This hasn't been vetted yet, since I'd like some feedback before putting time into writing tests. 

I think it'll work, but maybe we want to treat this more like the storage context, were we store the state in private property and then sync it periodically. We might have to see how this evolves as we actually use the models. Personally I like the idea of not having to worry about when to `load` and `persist` temporary state.

File: packages/fxa-settings/src/lib/context/implementations/url-context.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/context/implementations/url-context.ts:25]: We might need introduce some JSON serialization here. I'll address this later when tests are added.

File: packages/fxa-settings/src/models/reliers/rellier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/rellier.ts:99]: This holds the actual state of the component. All the `bind` decorators effectively bind the model to whatever state is held inside this context object. Currently there are two implementations one of URL query parameter contexts and one for local/session storage contexts. We could make the context's implementation as sophisticated as needed, but at the end of the day the model shouldn't really be concerned with how the data is stored / retrieved.

File: packages/fxa-settings/tsconfig.json
Comment (dschom) [inline packages/fxa-settings/tsconfig.json:4]: I know it says experiemental, but we are using these all over the place, for example nest relies on decorators. Looking at the JS code that gets generated decorators are actually pretty straight forward, and I've used them in production systems before without issue.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/context/context-validation.ts:5]: I went down a pretty big rabbit hole trying to port this, and decided I better check in with you all first. I think it's doable, but lack of types in VAT made it a pain...

File: packages/fxa-settings/src/lib/context/context-transforms.ts
Comment (dschom) [inline packages/fxa-settings/src/lib/context/context-transforms.ts:8]: We might need other transforms on the values as well. For example coercing encoded URI components to a typed object. I think this will be more obvious once tests are ported.

File: packages/fxa-settings/src/lib/hooks.tsx
Comment (dschom) [inline packages/fxa-settings/src/lib/hooks.tsx:64]: We could also maybe just do this setup as part of the AppContext... I didn't look to closely at since I know we were talking about using hooks.

File: packages/fxa-settings/src/models/reliers/rellier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/rellier.ts:109]: There might be some other convenience methods needed here as well. This will become more obvious when tests are ported.

File: packages/fxa-settings/src/models/reliers/rellier.ts
Comment (dschom) [inline packages/fxa-settings/src/models/reliers/rellier.ts:106]: This probably isn't needed, since the current behavior validates when data is written and read. However, we might want to change depending cost of validation. Also the first the context is loaded, it probably makes sense to validate the state and fail fast if something is off.

File: packages/fxa-settings/src/models/reliers/oauth-relier.ts
Comment (pdehaan) [inline packages/fxa-settings/src/models/reliers/oauth-relier.ts:7]: Typo in filename? "./rellier" => "./relier"?

File: packages/fxa-settings/src/lib/hooks.tsx
Comment (pdehaan) [inline packages/fxa-settings/src/lib/hooks.tsx:6]: filename should be "../models/reliers/relier"?

File: packages/fxa-settings/src/lib/context/interfaces/context-provider.ts
Comment (LZoog) [inline packages/fxa-settings/src/lib/context/interfaces/context-provider.ts:7]: It feels a little funny naming these `context` / `context-provider` to me since I think it may cause some confusion when creating or referencing React's context + providers but I'm not sure of better naming off the top of my head except maybe `model` for `context-provider`...

File: packages/fxa-settings/src/models/reliers/rellier.ts
Comment (LZoog) [inline packages/fxa-settings/src/models/reliers/rellier.ts:69]: FWIW we started an enum called `MozServices` that we'll add to as we confirm service names during development. Not sure if it should be used here or not since it's incomplete though.

Also, not sure if it's possible to reuse the types from `RelierData` but that could be nice. 🤔

File: packages/fxa-settings/src/models/reliers/rellier.ts
Comment (LZoog) [reply packages/fxa-settings/src/models/reliers/rellier.ts:99]: Could use a comment perhaps with your first 1-2 sentences. Thanks for the comments in-code and self-review btw.

File: packages/fxa-settings/src/models/reliers/options/browser-relier-options.ts
Comment (LZoog) [inline packages/fxa-settings/src/models/reliers/options/browser-relier-options.ts:7]: Hmm, maybe reference this (or make reference here) in whatever follow up ticket we make since yeah we should be able to nix this.

File: packages/fxa-settings/src/models/reliers/oauth-relier.ts
Comment (LZoog) [reply packages/fxa-settings/src/models/reliers/oauth-relier.ts:25]: Super nice 👌

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (LZoog) [inline packages/fxa-settings/src/lib/relier-factory.ts:16]: I reference the `MozServices` enum further down in review, this makes me wonder where we should account for 'sync' vs 'Firefox Sync' hmm.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (LZoog) [inline packages/fxa-settings/src/lib/relier-factory.ts:65]: Does this need to be `||` instead of `&&`? I see this comment:

```
...
      } else if (
        this._isServiceSync() ||
        // context v3 is able to sign in to Firefox without enabling Sync
        this._searchParam('context') === Constants.FX_DESKTOP_V3_CONTEXT
      ) {
...
```

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (LZoog) [inline packages/fxa-settings/src/lib/relier-factory.ts:129]: Is it still conventional to use underscores when the method is declared `private`? I guess I see a mix of both in the codebase.

**Review by LZoog — COMMENTED** [review]
Going to go ahead and submit my initial review here, it's mostly nits or just comments/questions. I think the overall pattern is there and this is awesome to see coming together. 🤘

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (LZoog) [reply packages/fxa-settings/src/lib/relier-factory.ts:174]: Are you wondering why we need to verify that browser is the same or why we're using it to verify (maybe there's a better way)?

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (vbudhram) [reply packages/fxa-settings/src/lib/context/context-validation.ts:5]: Basic validation is fine by me 👍🏽

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (vbudhram) [reply packages/fxa-settings/src/lib/relier-factory.ts:174]: I'm honestly not sure why this is? Possibly it was written to storage context so that it could continue the flow if continued in another tab/window. This might be one of these things we experiment with removing and see what breaks 😅

**Review by vbudhram — COMMENTED** [review]
@dschom The approach seems fine to me! Is there a minimum subset of models we can land with tests before adding them all? Maybe the base relier model?

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/relier-factory.ts:174]: @LZoog  I was mostly curious why we need to verify the browser is the same. I tried various user flows locally and never saw this state in my local storage. There's probably some use case that sets it, but I don't know what that is. This is used when we call isOauth(), and some of this logic in that function seems to overlap a bit, so maybe  this check isn't super necessary 🤷? (I bet it is though.)

@vbudhram Experimentation sounds good! Do you think maybe it has something to do with sync? If any experiments come to mind that'd trigger this state in storage, let me know.

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/storage-context.ts:26]: I just looked at the storage.ts class more closely. I like this a lot better and it should conform with Context interface, so I'll try switching to this. Thanks for the call out 👍. Probably no need for a follow up. If the namespace proves to a problem, I am sure it can worked around if necessary.

File: packages/fxa-settings/src/lib/context/implementations/url-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/url-context.ts:34]: Oh, yes replaceState, good call 😅.

File: packages/fxa-settings/src/lib/hooks.tsx
Comment (dschom) [reply packages/fxa-settings/src/lib/hooks.tsx:6]: Or, `../models/relier` . Thanks for the call out regardless. 🙇‍♂️

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/storage-context.ts:88]: I did experimented with this a little bit. I'll revisit it again though. As I recall, I ran into a problem. At one point I actually had a generic in place (ie `set<T>(key:string, value:T)`), because I thought I could propagate the type all the way into this storage layer. This didn't end up working, but I don't recall why anymore.

It's good to note that because we use these storage contexts as a back end store for the `@bind` decorator, an interesting thing happens. The `@bind` decorator is inherently typed, and it also has validation that should enforce its type. 

Therefore we actually retain type safety as long as we only access the storage context through fields/properties that have been decorated with the `@bind` directive. Obviously if we go behind the bind decorator, all bets are off. 

Anyway, since the @bind decorator is type safe, I feel a bit better about the `any` type here. This store is essentially just one small abstraction layer between the model and the storage api (or other kv store). The storage API can't strong types anyways, so the fact we have a general `any` here doesn't bother me too much.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/relier-factory.ts:16]: yeah... I am sure there are kinks with our state to work out. Hopefully testing shines some light on this. Also maybe these consts should go somewhere else?

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/relier-factory.ts:65]: Nice catch, thank you.

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/relier-factory.ts:129]: We've decided to mostly follow the google typescript guide, which says no. I literally copied and pasted 90% of this from app-start in content server. I'll clean it up once the tests settle.

File: packages/fxa-settings/src/lib/context/bind-decorator.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/bind-decorator.ts:89]: 🦖

**Comment by dschom** [comment]
> @dschom The approach seems fine to me! Is there a minimum subset of models we can land with tests before adding them all? Maybe the base relier model?

@vbudhram that's a good idea. I'll do another pass with a minimal set of tests, and let you see how this looks. Are you thinking that the relier would unblock some work?

**Comment by vbudhram** [comment]
> > @dschom The approach seems fine to me! Is there a minimum subset of models we can land with tests before adding them all? Maybe the base relier model?
> 
> @vbudhram that's a good idea. I'll do another pass with a minimal set of tests, and let you see how this looks. Are you thinking that the relier would unblock some work?

Yea, I could probably use the base relier in the password reset tickets. It doesn't block that work though, so if it is alot of overhead I would keep it together. Your call!

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (vbudhram) [inline packages/fxa-settings/src/lib/context/context-validation.ts:5]: I think the remaining bits here can be fixed in a follow up.

**Review by vbudhram — COMMENTED** [review]
@dschom This matches what we talked about in slack and LGTM. I would be in favor of filing follow ups to finish up remaining test coverage. I'm fairly certain I missed some stuff in this review but it isn't anything that we couldn't fix up while integrating it.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:5]: @vbudhram I got some basic stuff in there. I think we should file a follow up for testing it thoroughly and flushing out all the different validation types. I did my best add them as they were encountered.

File: packages/fxa-settings/src/lib/hooks.tsx
Comment (dschom) [reply packages/fxa-settings/src/lib/hooks.tsx]: @LZoog I think this is pretty clean now. The relier factory is added to the app context, and then there is a small hook that makes it easy to get the current relier from that factory via `useRelier`.

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/storage-context.ts:94]: @LZoog I think we need a follow up for the storage context thing. I'd like to use the storage class that exists in fxa-settings since it seems quite a bit better, but I think that will require migrating the existing models... or modifying how the namespace thing works. This was mentioned in some other places in the review as well, but I'll bring it up again here. Do you want me to file a ticket for this?

File: packages/fxa-settings/src/lib/context/interfaces/context-provider.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/interfaces/context-provider.ts:7]: Lol, good call out. This name keeps confusing VSCODE auto imports. I'll come up with something better.

Comment (dschom) [reply packages/fxa-settings/src/lib/context/interfaces/context-provider.ts:7]: Okay, I am thinking make this more explicit. 

`context-provider -> BoundModel` _(It has a validate method so is more than a pure provider now anyways.)_

`context -> ModelContext` _(Which still conflicts a little bit, but I think at least pretty explicit)_

**Comment by dschom** [comment]
> @dschom This matches what we talked about in slack and LGTM. I would be in favor of filing follow ups to finish up remaining test coverage. I'm fairly certain I missed some stuff in this review but it isn't anything that we couldn't fix up while integrating it.

Agreed, I'll start filing follow ups.

File: packages/fxa-settings/src/models/reliers/signin-signup-info.ts
Comment (dschom) [reply packages/fxa-settings/src/models/reliers/signin-signup-info.ts:56]: I think having a validator for `isUri` make sense. 

`redirectTo`, `redirectUri` and `redirectUrl` (Used by oauth models) are valid field names and all seem to have the same purpose. I guess I am not sure if it was intentional to keep them separate, or if this was just the way things evolved. I'll file a follow up for this.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:85]: Thanks for calling it out. I didn't spend too much effort on the validators since we have a follow up issue to create tests and flush them out, none the less. But you're right, this doesn't make much sense though! I'll clean it up a bit.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:81]: I should be consistent. Thanks for the call out.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:156]: I think I'll just change it to `isGreaterThanZero` to be explicit.

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/storage-context.ts:29]: @LZoog I ended up filing a follow up for this. I started to look into it but don't want to miss the train. See FXA-6801.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:5]: @vbudhram A follow up issue has been filed. See FXA-6800.

File: packages/fxa-settings/src/lib/context/implementations/storage-context.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/implementations/storage-context.ts:94]: Follow up created: FXA-6801

File: packages/fxa-settings/src/lib/relier-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/relier-factory.ts]: I'll have to file a follow up for this. I tried to change this, but once I made the switch, there were quite a few places that broke. It seems like we have a handful of hard coded strings that are anticipated or even set as defaults which don't align with MozServices. See FXA-6802.

File: packages/fxa-settings/src/lib/context/context-validation.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/context/context-validation.ts:5]: Issue filed: FXA-6800

File: packages/fxa-settings/src/models/reliers/signin-signup-info.ts
Comment (dschom) [reply packages/fxa-settings/src/models/reliers/signin-signup-info.ts:56]: There are now both isUri and isUrl validators. For now, we will use isUrl unless the field explicitly indicates a URI.

**Review by vbudhram — APPROVED** [review]
@dschom I think we can go ahead and land this since you have filed follow ups. Thanks!

**Pull request merged**
